File: app/src/main/kotlin/org/terst/cancan/cooking/CookingScreen.kt
package org.terst.cancan.cooking

import android.app.Activity
import android.view.WindowManager
import androidx.compose.animation.animateColorAsState
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.MenuBook
import androidx.compose.material.icons.filled.Pause
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.Replay
import androidx.compose.material.icons.filled.Timer
import androidx.compose.material3.Button
import androidx.compose.material3.ElevatedCard
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FilledTonalButton
import androidx.compose.material3.FilledTonalIconButton
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.ListItem
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.navigation.NavController
import org.terst.cancan.recipes.data.RecipeItem

private val TIMER_PRESETS = listOf(5, 10, 15, 20, 30, 45, 60)

@Composable
fun CookingScreen(navController: NavController) {
    val viewModel: CookingViewModel = hiltViewModel()
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()
    val activity = LocalContext.current as? Activity

    LaunchedEffect(uiState.isActive) {
        if (uiState.isActive) {
            activity?.window?.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        } else {
            activity?.window?.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
    }
    DisposableEffect(Unit) {
        onDispose { activity?.window?.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON) }
    }

    LazyColumn(
        contentPadding = PaddingValues(horizontal = 16.dp, vertical = 4.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
        modifier = Modifier.fillMaxSize()
    ) {
        item {
            Text(
                text = "Cooking",
                style = MaterialTheme.typography.headlineMedium,
                modifier = Modifier.padding(vertical = 8.dp)
            )
        }

        uiState.session?.let { session ->
            item {
                SessionCard(
                    session = session,
                    onPrev = viewModel::prevStep,
                    onNext = viewModel::nextStep,
                    onEnd = viewModel::endSession,
                    onAddTimer = viewModel::addTimer
                )
            }
        }

        item {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth()
            ) {
                Text(
                    text = "Timers",
                    style = MaterialTheme.typography.titleMedium,
                    modifier = Modifier.weight(1f)
                )
                FilledTonalButton(onClick = viewModel::showAddTimer) {
                    Icon(Icons.Default.Add, contentDescription = null, modifier = Modifier.size(18.dp))
                    Spacer(Modifier.width(4.dp))
                    Text("Add Timer")
                }
            }
        }

        if (uiState.timers.isEmpty()) {
            item {
                Text(
                    text = "No active timers.\nTap \"Add Timer\" or start a recipe.",
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 16.dp),
                    textAlign = TextAlign.Center
                )
            }
        } else {
            items(uiState.timers, key = { it.id }) { timer ->
                TimerCard(
                    timer = timer,
                    onToggle = { viewModel.toggleTimer(timer.id) },
                    onReset = { viewModel.resetTimer(timer.id) },
                    onRemove = { viewModel.removeTimer(timer.id) }
                )
            }
        }

        if (uiState.session == null) {
            item {
                OutlinedButton(
                    onClick = viewModel::showRecipePicker,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(bottom = 8.dp)
                ) {
                    Icon(Icons.Default.MenuBook, contentDescription = null, modifier = Modifier.size(18.dp))
                    Spacer(Modifier.width(8.dp))
                    Text("Follow a Recipe")
                }
            }
        }
    }

    if (uiState.showAddTimer) {
        AddTimerSheet(onAdd = viewModel::addTimer, onDismiss = viewModel::dismissAddTimer)
    }

    if (uiState.showRecipePicker) {
        RecipePickerSheet(
            recipes = uiState.recipes,
            onSelect = viewModel::startSession,
            onDismiss = viewModel::dismissRecipePicker
        )
    }
}

@Composable
private fun SessionCard(
    session: CookingSession,
    onPrev: () -> Unit,
    onNext: () -> Unit,
    onEnd: () -> Unit,
    onAddTimer: (String, Long) -> Unit
) {
    val step = session.currentStep ?: return
    val suggestedMs = remember(step.instruction) { parseDurationMs(step.instruction) }

    ElevatedCard(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = session.recipe.name,
                        style = MaterialTheme.typography.titleSmall,
                        color = MaterialTheme.colorScheme.primary
                    )
                    Text(
                        text = "Step ${session.stepIndex + 1} of ${session.stepCount}",
                        style = MaterialTheme.typography.labelMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
                TextButton(onClick = onEnd) { Text("End") }
            }

            LinearProgressIndicator(
                progress = { (session.stepIndex + 1f) / session.stepCount },
                modifier = Modifier.fillMaxWidth()
            )

            Text(text = step.instruction, style = MaterialTheme.typography.bodyLarge)

            step.tips.forEach { tip ->
                Text(
                    text = "Tip: $tip",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }

            if (suggestedMs != null) {
                FilledTonalButton(
                    onClick = {
                        onAddTimer(
                            "${session.recipe.name} – step ${session.stepIndex + 1}",
                            suggestedMs
                        )
                    },
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Icon(Icons.Default.Timer, contentDescription = null, modifier = Modifier.size(18.dp))
                    Spacer(Modifier.width(4.dp))
                    Text("Start ${formatMs(suggestedMs)} timer")
                }
            }

            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                OutlinedButton(
                    onClick = onPrev,
                    enabled = !session.isFirst,
                    modifier = Modifier.weight(1f)
                ) { Text("← Prev") }
                Button(
                    onClick = onNext,
                    enabled = !session.isLast,
                    modifier = Modifier.weight(1f)
                ) { Text("Next →") }
            }
        }
    }
}

@Composable
private fun TimerCard(
    timer: TimerItem,
    onToggle: () -> Unit,
    onReset: () -> Unit,
    onRemove: () -> Unit
) {
    val timeColor by animateColorAsState(
        targetValue = when {
            timer.isFinished -> MaterialTheme.colorScheme.error
            timer.isRunning && timer.remainingMs < 60_000L -> MaterialTheme.colorScheme.tertiary
            else -> MaterialTheme.colorScheme.onSurface
        },
        label = "timerColor"
    )

    ElevatedCard(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier.padding(horizontal = 16.dp, vertical = 12.dp),
            verticalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = timer.name,
                    style = MaterialTheme.typography.titleSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.weight(1f)
                )
                IconButton(onClick = onRemove, modifier = Modifier.size(32.dp)) {
                    Icon(
                        Icons.Default.Close,
                        contentDescription = "Remove timer",
                        modifier = Modifier.size(16.dp)
                    )
                }
            }

            Text(
                text = formatMs(timer.remainingMs),
                style = MaterialTheme.typography.displayMedium,
                fontFamily = FontFamily.Monospace,
                color = timeColor,
                modifier = Modifier.fillMaxWidth(),
                textAlign = TextAlign.Center
            )

            if (timer.isFinished) {
                Text(
                    text = "Done!",
                    style = MaterialTheme.typography.labelLarge,
                    color = MaterialTheme.colorScheme.error,
                    modifier = Modifier.fillMaxWidth(),
                    textAlign = TextAlign.Center
                )
            } else {
                LinearProgressIndicator(
                    progress = { timer.progress },
                    modifier = Modifier.fillMaxWidth()
                )
            }

            Row(
                horizontalArrangement = Arrangement.Center,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 4.dp)
            ) {
                if (!timer.isFinished) {
                    FilledTonalIconButton(onClick = onToggle) {
                        Icon(
                            imageVector = if (timer.isRunning) Icons.Default.Pause else Icons.Default.PlayArrow,
                            contentDescription = if (timer.isRunning) "Pause" else "Resume"
                        )
                    }
                    Spacer(Modifier.width(8.dp))
                }
                FilledTonalIconButton(onClick = onReset) {
                    Icon(Icons.Default.Replay, contentDescription = "Reset timer")
                }
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun AddTimerSheet(onAdd: (String, Long) -> Unit, onDismiss: () -> Unit) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    var name by rememberSaveable { mutableStateOf("") }
    var selectedPreset by rememberSaveable { mutableStateOf(10) }
    var customStr by rememberSaveable { mutableStateOf("") }
    var useCustom by rememberSaveable { mutableStateOf(false) }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        modifier = Modifier.imePadding()
    ) {
        Column(
            modifier = Modifier
                .padding(horizontal = 16.dp)
                .padding(bottom = 24.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            Text("Add Timer", style = MaterialTheme.typography.titleLarge)

            OutlinedTextField(
                value = name,
                onValueChange = { name = it },
                label = { Text("Name (optional)") },
                placeholder = { Text("e.g. Processing time") },
                singleLine = true,
                modifier = Modifier.fillMaxWidth()
            )

            Text(
                text = "Duration",
                style = MaterialTheme.typography.labelLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )

            LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                items(TIMER_PRESETS) { min ->
                    FilterChip(
                        selected = !useCustom && selectedPreset == min,
                        onClick = { selectedPreset = min; useCustom = false },
                        label = { Text("${min}m") }
                    )
                }
                item {
                    FilterChip(
                        selected = useCustom,
                        onClick = { useCustom = true },
                        label = { Text("Custom") }
                    )
                }
            }

            if (useCustom) {
                OutlinedTextField(
                    value = customStr,
                    onValueChange = { customStr = it.filter(Char::isDigit) },
                    label = { Text("Minutes") },
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                    singleLine = true,
                    modifier = Modifier.fillMaxWidth()
                )
            }

            val durationMs = if (useCustom) {
                (customStr.toLongOrNull() ?: 0L) * 60_000L
            } else {
                selectedPreset * 60_000L
            }

            Button(
                onClick = { onAdd(name.trim(), durationMs) },
                enabled = durationMs > 0,
                modifier = Modifier.fillMaxWidth()
            ) {
                Icon(Icons.Default.Timer, contentDescription = null, modifier = Modifier.size(18.dp))
                Spacer(Modifier.width(8.dp))
                Text("Start Timer")
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun RecipePickerSheet(
    recipes: List<RecipeItem>,
    onSelect: (RecipeItem) -> Unit,
    onDismiss: () -> Unit
) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = sheetState) {
        Column(modifier = Modifier.padding(horizontal = 16.dp)) {
            Text(
                text = "Choose a Recipe",
                style = MaterialTheme.typography.titleLarge,
                modifier = Modifier.padding(bottom = 8.dp)
            )
            LazyColumn(
                contentPadding = PaddingValues(bottom = 24.dp),
                verticalArrangement = Arrangement.spacedBy(2.dp),
                modifier = Modifier.heightIn(max = 440.dp)
            ) {
                items(recipes) { recipe ->
                    ListItem(
                        headlineContent = { Text(recipe.name) },
                        supportingContent = {
                            val steps = "${recipe.steps.size} steps"
                            val meta = if (recipe.timeEstimate.isNotBlank()) {
                                "${recipe.category} · $steps · ${recipe.timeEstimate}"
                            } else {
                                "${recipe.category} · $steps"
                            }
                            Text(meta)
                        },
                        modifier = Modifier.clickable { onSelect(recipe) }
                    )
                }
            }
        }
    }
}

private fun formatMs(ms: Long): String {
    val total = (ms / 1000).coerceAtLeast(0L)
    val h = total / 3600
    val m = (total % 3600) / 60
    val s = total % 60
    return if (h > 0) "%d:%02d:%02d".format(h, m, s) else "%02d:%02d".format(m, s)
}

private fun parseDurationMs(text: String): Long? {
    val regex = Regex("""(\d+)\s*(?:to \d+ )?minutes?""", RegexOption.IGNORE_CASE)
    val match = regex.find(text) ?: return null
    return match.groupValues[1].toLong() * 60_000L
}

File: flutter/lib/features/cooking/cooking_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:wakelock_plus/wakelock_plus.dart';
import '../../data/models/recipe_item.dart';
import '../recipes/recipes_notifier.dart';
import 'cooking_notifier.dart';

class CookingScreen extends ConsumerStatefulWidget {
  const CookingScreen({super.key});

  @override
  ConsumerState<CookingScreen> createState() => _CookingScreenState();
}

class _CookingScreenState extends ConsumerState<CookingScreen> {
  @override
  void dispose() {
    WakelockPlus.disable();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final cookState = ref.watch(cookingProvider);
    final hasActivity =
        cookState.session != null || cookState.timers.any((t) => t.running);

    if (hasActivity) {
      WakelockPlus.enable();
    } else {
      WakelockPlus.disable();
    }

    return Scaffold(
      appBar: AppBar(title: const Text('Cooking')),
      body: ListView(
        padding: const EdgeInsets.all(12),
        children: [
          if (cookState.session != null)
            _SessionCard(session: cookState.session!),
          if (cookState.session == null)
            _StartSessionCard(),
          const SizedBox(height: 16),
          ...cookState.timers.map((t) => _TimerCard(timer: t)),
          OutlinedButton.icon(
            icon: const Icon(Icons.timer_outlined),
            label: const Text('Add Timer'),
            onPressed: () => _showAddTimer(context),
          ),
        ],
      ),
    );
  }

  void _showAddTimer(BuildContext context) {
    showModalBottomSheet(
      context: context,
      isScrollControlled: true,
      builder: (_) => _AddTimerSheet(),
    );
  }
}

class _StartSessionCard extends ConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final recipes = ref.watch(recipesProvider);
    return Card(
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            const Text('No active recipe session.',
                style: TextStyle(fontSize: 16)),
            const SizedBox(height: 12),
            FilledButton.icon(
              icon: const Icon(Icons.restaurant_menu),
              label: const Text('Start Recipe'),
              onPressed: () => recipes.whenData((list) =>
                  _showRecipePicker(context, ref, list)),
            ),
          ],
        ),
      ),
    );
  }

  void _showRecipePicker(
      BuildContext context, WidgetRef ref, List<RecipeItem> recipes) {
    showModalBottomSheet(
      context: context,
      isScrollControlled: true,
      builder: (_) => DraggableScrollableSheet(
        expand: false,
        initialChildSize: 0.6,
        maxChildSize: 0.9,
        builder: (_, ctrl) => ListView.builder(
          controller: ctrl,
          itemCount: recipes.length,
          itemBuilder: (_, i) => ListTile(
            title: Text(recipes[i].name),
            subtitle: Text(recipes[i].category),
            onTap: () {
              ref.read(cookingProvider.notifier).startSession(recipes[i]);
              Navigator.pop(context);
            },
          ),
        ),
      ),
    );
  }
}

class _SessionCard extends ConsumerWidget {
  final CookingSession session;
  const _SessionCard({required this.session});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final notifier = ref.read(cookingProvider.notifier);
    final theme = Theme.of(context);
    final step = session.step;
    final progress = (session.currentStep + 1) / session.recipe.steps.length;

    final timerMatch = RegExp(r'(\d+)\s*minute').firstMatch(step.instruction);

    return Card(
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Row(children: [
              Expanded(
                child: Text(session.recipe.name,
                    style: theme.textTheme.titleMedium
                        ?.copyWith(fontWeight: FontWeight.bold)),
              ),
              TextButton(
                  onPressed: notifier.endSession,
                  child: const Text('End')),
            ]),
            const SizedBox(height: 4),
            Text(
                'Step ${session.currentStep + 1} of ${session.recipe.steps.length}',
                style: theme.textTheme.bodySmall),
            const SizedBox(height: 6),
            LinearProgressIndicator(value: progress),
            const SizedBox(height: 12),
            Text(step.instruction, style: theme.textTheme.bodyLarge),
            ...step.tips.map(
              (tip) => Padding(
                padding: const EdgeInsets.only(top: 6),
                child: Text('💡 $tip',
                    style: theme.textTheme.bodySmall),
              ),
            ),
            if (timerMatch != null) ...[
              const SizedBox(height: 8),
              OutlinedButton.icon(
                icon: const Icon(Icons.timer),
                label: Text('Start ${timerMatch.group(1)} min timer'),
                onPressed: () => ref
                    .read(cookingProvider.notifier)
                    .addTimer('Step ${session.currentStep + 1}',
                        Duration(minutes: int.parse(timerMatch.group(1)!))),
              ),
            ],
            const SizedBox(height: 12),
            Row(children: [
              Expanded(
                child: OutlinedButton(
                  onPressed: session.isFirst ? null : notifier.prevStep,
                  child: const Text('← Prev'),
                ),
              ),
              const SizedBox(width: 8),
              Expanded(
                child: FilledButton(
                  onPressed: session.isLast ? null : notifier.nextStep,
                  child: const Text('Next →'),
                ),
              ),
            ]),
          ],
        ),
      ),
    );
  }
}

class _TimerCard extends ConsumerWidget {
  final TimerItem timer;
  const _TimerCard({required this.timer});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final notifier = ref.read(cookingProvider.notifier);
    final theme = Theme.of(context);
    final remaining = timer.remaining;
    final progress = timer.total.inSeconds == 0
        ? 0.0
        : (timer.total.inSeconds - remaining.inSeconds) /
            timer.total.inSeconds;

    Color timeColor = theme.colorScheme.onSurface;
    if (timer.isDone) {
      timeColor = theme.colorScheme.primary;
    } else if (remaining.inSeconds < 60) {
      timeColor = theme.colorScheme.error;
    } else if (timer.running) {
      timeColor = theme.colorScheme.tertiary;
    }

    String timeStr;
    if (remaining.inHours > 0) {
      timeStr =
          '${remaining.inHours}:${(remaining.inMinutes % 60).toString().padLeft(2, '0')}:${(remaining.inSeconds % 60).toString().padLeft(2, '0')}';
    } else {
      timeStr =
          '${remaining.inMinutes.toString().padLeft(2, '0')}:${(remaining.inSeconds % 60).toString().padLeft(2, '0')}';
    }

    return Card(
      margin: const EdgeInsets.only(bottom: 8),
      child: Padding(
        padding: const EdgeInsets.all(12),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Row(children: [
              Expanded(
                  child: Text(timer.name.isNotEmpty ? timer.name : 'Timer',
                      style: const TextStyle(fontWeight: FontWeight.w600))),
              IconButton(
                icon: const Icon(Icons.close, size: 18),
                onPressed: () => notifier.removeTimer(timer.id),
              ),
            ]),
            Text(
              timer.isDone ? 'Done!' : timeStr,
              style: theme.textTheme.displaySmall
                  ?.copyWith(color: timeColor, fontWeight: FontWeight.bold),
              textAlign: TextAlign.center,
            ),
            const SizedBox(height: 6),
            LinearProgressIndicator(value: progress),
            const SizedBox(height: 8),
            Row(children: [
              Expanded(
                child: FilledButton.icon(
                  icon: Icon(timer.running ? Icons.pause : Icons.play_arrow),
                  label: Text(timer.running ? 'Pause' : 'Start'),
                  onPressed:
                      timer.isDone ? null : () => notifier.toggleTimer(timer.id),
                ),
              ),
              const SizedBox(width: 8),
              OutlinedButton(
                onPressed: () => notifier.resetTimer(timer.id),
                child: const Text('Reset'),
              ),
            ]),
          ],
        ),
      ),
    );
  }
}

class _AddTimerSheet extends ConsumerStatefulWidget {
  @override
  ConsumerState<_AddTimerSheet> createState() => _AddTimerSheetState();
}

class _AddTimerSheetState extends ConsumerState<_AddTimerSheet> {
  final _nameCtrl = TextEditingController();
  final _customCtrl = TextEditingController();
  int? _presetMinutes;

  static const presets = [5, 10, 15, 20, 30, 45, 60];

  @override
  void dispose() {
    _nameCtrl.dispose();
    _customCtrl.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.only(
        left: 16, right: 16, top: 16,
        bottom: MediaQuery.of(context).viewInsets.bottom + 16,
      ),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Text('Add Timer', style: Theme.of(context).textTheme.titleLarge),
          const SizedBox(height: 12),
          TextField(
            controller: _nameCtrl,
            decoration: const InputDecoration(
                labelText: 'Name (optional)', border: OutlineInputBorder()),
          ),
          const SizedBox(height: 12),
          Wrap(
            spacing: 8,
            children: presets
                .map(
                  (m) => ChoiceChip(
                    label: Text('$m min'),
                    selected: _presetMinutes == m,
                    onSelected: (_) =>
                        setState(() => _presetMinutes = m == _presetMinutes ? null : m),
                  ),
                )
                .toList(),
          ),
          const SizedBox(height: 8),
          TextField(
            controller: _customCtrl,
            keyboardType: TextInputType.number,
            decoration: const InputDecoration(
                labelText: 'Custom minutes', border: OutlineInputBorder()),
            onChanged: (_) => setState(() => _presetMinutes = null),
          ),
          const SizedBox(height: 16),
          FilledButton(
            onPressed: _canAdd ? _add : null,
            child: const Text('Add Timer'),
          ),
        ],
      ),
    );
  }

  bool get _canAdd =>
      _presetMinutes != null || (int.tryParse(_customCtrl.text) ?? 0) > 0;

  void _add() {
    final minutes =
        _presetMinutes ?? int.tryParse(_customCtrl.text) ?? 0;
    if (minutes <= 0) return;
    ref.read(cookingProvider.notifier).addTimer(
          _nameCtrl.text.trim(),
          Duration(minutes: minutes),
        );
    Navigator.pop(context);
  }
}
